src/Database/LSMTree/Internal/Run/Index/Compact.hs

```diff
@@ -1,21 +1,506 @@
+-------------------------------------------------------------------------------}
+
+-- | Inclusive bounds for range-finder bit-precision.
+rangeFinderPrecisionBounds :: (Int, Int)
+rangeFinderPrecisionBounds = (0, 16)
+
+-- | https://en.wikipedia.org/wiki/Birthday_problem#Probability_of_a_shared_birthday_(collision)
+suggestRangeFinderPrecision :: Int -> Int
+suggestRangeFinderPrecision maxPages =
+    clamp $ ceiling @Double $ 2 * logBase 2 (fromIntegral maxPages)
+  where
+    (lb, ub)            = rangeFinderPrecisionBounds
+    clamp x | x < lb    = lb
+            | x > ub    = ub
+            | otherwise = x
+
+{-------------------------------------------------------------------------------
+  Queries
+-------------------------------------------------------------------------------}
+
+-- | Given a search key, find the number of the disk page that /might/ contain
+-- this key.
+--
+-- Note: there is no guarantee that the disk page contains the key. However, it
+-- is guaranteed that /no other/ disk page contain the key.
+--
+search ::
+     forall k. (SliceBits k, Integral k)
+  => k
+  -> CompactIndex k
+  -> Maybe Int
+search k CompactIndex{..} = runST $ do
+    let lb = ciRangeFinder VU.! fromIntegral (topBits ciRangeFinderPrecision k)
+        ub = ciRangeFinder VU.! fromIntegral (topBits ciRangeFinderPrecision k + 1)
+    -- We have to thaw the primary index to use the vector search algorithm. It is
+    -- safe to use 'unsafeThaw' here because, once frozen, a compact index can not
+    -- be manipulated anymore.
+    !ciPrimary' <- VU.unsafeThaw ciPrimary
+    -- Compute @pageNr@ similar to 'IntMap.lookupLE': find the page number
+    -- @pageNr@ of the the last entry in the primary array (within bounds @[lb,
+    -- ub]@) that is smaller or equal to our search key.
+    --
+    -- Regardless of whether the search key is actually present in a page, it
+    -- /can not/ exist in a page @>pageNr@.
+    pageNr' <- VA.gallopingSearchLeftPBounds
+                (> sliceBits32 ciRangeFinderPrecision 32 k)
+                ciPrimary'
+                (fromIntegral lb)
+                (fromIntegral ub)
+    let pageNr = pageNr' - 1
+    if pageNr < 0 then
+      -- The search key is smaller than the minimum key of the first page.
+      pure Nothing
+    else
+      -- Consult the tie-breaker index, if necessary.
+      pure $ breakTies pageNr (unBit $ ciClashes VU.! pageNr)
+  where
+    breakTies :: Int -> Bool -> Maybe Int
+    breakTies pageNr clash
+      -- Clash! The page we're currently looking at has a full minimum key that
+      -- is smaller than our search key, so we're done.
+      | clash
+      , ciTieBreaker IntMap.! pageNr <= k
+      = Just pageNr
+      -- Clash! The page we're currently looking at has a full minimum key that
+      -- is strictly larger than our search key, so we have to look at the page
+      -- before it.
+      | clash
+      = breakTies (pageNr - 1) (unBit $ ciClashes VU.! (pageNr - 1))
```

Comment:
Hmm, for efficiency, do we have to make this recursive? Doesn't it only go round once more here?